paddle/fluid/operators/data_norm_op.cu

```diff
@@ -0,0 +1,232 @@
+    const float dr = ctx.Attr<float>("summary_decay_rate");
+    const int sync_with_nccl = ctx.Attr<bool>("sync_with_nccl");
+
+    const auto &x_dims = x->dims();
+    // Align with CPU version, but should we add this restriction?
+    PADDLE_ENFORCE_EQ(x_dims.size(), 2, platform::errors::PreconditionNotMet(
+                                            "The Input dim size should be 2"));
+    const int N = x_dims[0];
+    const int C = x_dims[1];
+
+    // init output
+    Tensor *d_x = nullptr;
+    if (ctx.HasOutput(framework::GradVarName("X"))) {
+      d_x = ctx.Output<Tensor>(framework::GradVarName("X"));
+    }
+    T *d_batch_size = ctx.Output<Tensor>(framework::GradVarName("BatchSize"))
+                          ->mutable_data<T>(ctx.GetPlace());
+    T *d_batch_sum = ctx.Output<Tensor>(framework::GradVarName("BatchSum"))
+                         ->mutable_data<T>(ctx.GetPlace());
+    T *d_batch_square_sum =
+        ctx.Output<Tensor>(framework::GradVarName("BatchSquareSum"))
+            ->mutable_data<T>(ctx.GetPlace());
+
+    auto stream =
+        ctx.template device_context<platform::CUDADeviceContext>().stream();
+    if (d_x != nullptr) {
+      kernel_data_norm_bp<<<GET_BLOCKS(C * N), PADDLE_CUDA_NUM_THREADS, 0,
+                            stream>>>(N, C, d_y->data<T>(), scales->data<T>(),
+                                      d_x->mutable_data<T>(ctx.GetPlace()));
+    }
+
+    kernel_data_norm_bp_stat<<<GET_BLOCKS(C), PADDLE_CUDA_NUM_THREADS, 0,
+                               stream>>>(N, C, x->data<T>(), means->data<T>(),
+                                         epsilon, d_batch_size, d_batch_sum,
+                                         d_batch_square_sum);
+
+    if (sync_with_nccl) {
+      auto comm = platform::NCCLCommContext::Instance().Get(0, ctx.GetPlace());
+      cudaStream_t comm_stream = comm->stream();
+
+      cudaError_t e_sync = cudaStreamSynchronize(stream);
+      if (e_sync != 0) {
+        LOG(FATAL) << "Fail to sync nccl stream: "
+                   << cudaGetErrorString(e_sync);
+      }
+      PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::ncclAllReduce(
+          reinterpret_cast<const void *>(d_batch_size),
+          reinterpret_cast<void *>(d_batch_size), C,
+          platform::ToNCCLDataType(x->type()), ncclSum, comm->comm(),
+          comm_stream));
+      PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::ncclAllReduce(
+          reinterpret_cast<const void *>(d_batch_sum),
+          reinterpret_cast<void *>(d_batch_sum), C,
+          platform::ToNCCLDataType(x->type()), ncclSum, comm->comm(),
+          comm_stream));
+      PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::ncclAllReduce(
+          reinterpret_cast<const void *>(d_batch_square_sum),
+          reinterpret_cast<void *>(d_batch_square_sum), C,
+          platform::ToNCCLDataType(x->type()), ncclSum, comm->comm(),
+          comm_stream));
+      e_sync = cudaStreamSynchronize(comm_stream);
```

Comment:
as synchronization is required, you would better to use calculation stream without call synchronizing.

Reply:
I think so, thanks